redis/src/cluster_async/mod.rs

```diff
@@ -700,28 +729,105 @@ where
     }
 
     // Query a node to discover slot-> master mappings with retries
-    fn refresh_slots_with_retries(&mut self) -> impl Future<Output = RedisResult<()>> {
-        let inner = self.inner.clone();
-        async move {
+    async fn refresh_slots_with_retries(inner: Arc<InnerCore<C>>) -> RedisResult<()> {
+        if inner
+            .slot_refresh_in_progress
+            .compare_exchange(false, true, Ordering::Acquire, Ordering::Relaxed)
+            .is_err()
+        {
+            return Ok(());
+        }
+        let retry_strategy = ExponentialBackoff {
+            initial_interval: DEFAULT_REFRESH_SLOTS_RETRY_INITIAL_INTERVAL,
+            max_interval: DEFAULT_REFRESH_SLOTS_RETRY_TIMEOUT,
+            ..Default::default()
+        };
+        let retries_counter = AtomicUsize::new(0);
+        let res = retry(retry_strategy, || {
+            retries_counter.fetch_add(1, atomic::Ordering::Relaxed);
+            Self::refresh_slots(
+                inner.clone(),
+                retries_counter.load(atomic::Ordering::Relaxed),
+            )
+            .map_err(Error::from)
+        })
+        .await;
+        // Set the slot_refresh_in_progress flag back to false
+        inner
+            .slot_refresh_in_progress
+            .store(false, Ordering::Relaxed);
+        res
+    }
+
+    async fn periodic_topology_check(
+        inner: Arc<InnerCore<C>>,
+        interval_duration: Duration,
+        shutdown_flag: Arc<AtomicBool>,
+    ) {
+        loop {
+            if shutdown_flag.load(Ordering::Acquire) {
+                return;
+            }
+            #[cfg(feature = "tokio-comp")]
+            tokio::time::sleep(interval_duration).await;
+            #[cfg(all(not(feature = "tokio-comp"), feature = "async-std-comp"))]
+            async_std::task::sleep(interval_duration).await;
+            let read_guard = inner.conn_lock.read().await;
+            if read_guard.get_current_topology_hash().is_none() {
```

Comment:
Actually, it cannot be None. When a new ClusterConnInner is being created, before we start the periodic checks we do refresh slots. If refresh slots failed it will fail the ClusterConnInner creation, otherwise it has to set the topology hash with Some(val). 
Do you think it's better if i'll remove the Option and initialize it with 0?

Reply:
sure, use 0.